geomesa-index-api/src/test/scala/org/locationtech/geomesa/index/utils/SortingSimpleFeatureIteratorTest.scala

```diff
@@ -0,0 +1,157 @@
+  val c2 = ScalaSimpleFeature.create(sft, "4", 9, "C")
+  val d =  ScalaSimpleFeature.create(sft, "5", 6, "D")
+
+  "SortingSimpleFeatureIterator" should {
+
+    "lazily sort" >> {
+
+      "when hasNext is called first" >> {
+        val features = mock[CloseableIterator[SimpleFeature]]
+        features.hasNext returns true thenReturns true thenReturns false
+        features.next returns b thenReturns a thenThrows new NoSuchElementException
+
+        val test = new SortingSimpleFeatureIterator(features, Seq(("", false)))
+
+        there were no(features).hasNext
+        there were no(features).next
+        there were no(features).close
+
+        test.hasNext must beTrue
+
+        there were three(features).hasNext
+        there were two(features).next
+        there were no(features).close
+
+        test.close()
+        there was one(features).close
+      }
+
+      "or when next is called first" >> {
+        val features = mock[CloseableIterator[SimpleFeature]]
+        features.hasNext returns true thenReturns true thenReturns false
+        features.next returns b thenReturns a thenThrows new NoSuchElementException
+
+        val test = new SortingSimpleFeatureIterator(features, Seq(("", false)))
+
+        there were no(features).hasNext
+        there were no(features).next
+        there were no(features).close
+
+        test.next mustEqual a
+
+        there were three(features).hasNext
+        there were two(features).next
+        there were no(features).close
+
+        test.close()
+        there was one(features).close
+      }
+    }
+
+    "be able to sort by id asc" >> {
+      val features = new TestCloseableIterator(Iterator(b, c1, d, a, c2))
+      test(features, Seq(("", false)), Seq(a, b, c1, c2, d))
+    }
+
+    "be able to sort by id desc" >> {
+      val features = new TestCloseableIterator(Iterator(b, c1, d, a, c2))
+      test(features, Seq(("", true)), Seq(d, c2, c1, b, a))
+    }
+
+    "be able to sort by an attribute asc" >> {
+      val features = new TestCloseableIterator(Iterator(b, c2, d, a, c1))
+      test(features, Seq(("name", false)), Seq(a, b, c2, c1, d)) // note: sort is stable
+    }
+
+    "be able to sort by an attribute desc" >> {
+      val features = new TestCloseableIterator(Iterator(b, c2, d, a, c1))
+      test(features, Seq(("name", true)), Seq(d, c2, c1, b, a)) // note: sort is stable
+    }
+
+    "be able to sort by an attribute and id" >> {
+      val features = new TestCloseableIterator(Iterator(b, c2, d, a, c1))
+      test(features, Seq(("name", false), ("", false)), Seq(a, b, c1, c2, d))
+    }
+
+    "be able to sort by an multiple attributes" >> {
+      val features = new TestCloseableIterator(Iterator(a, b, c1, d, c2))
+      test(features, Seq(("age", true), ("name", false)), Seq(b, c2, a, c1, d))
+    }
+
+    "be able to spill to disk" >> {
+      val threshold = Seq(a, b, c1, c2, d).map(_.calculateSizeOf()).min * 2
```

Comment:
How many files are you seeing this create on disk?

Reply:
i think it created 2 files. i couldn't think of a way to check that it was spilling to disk as a unit test assertion without exposing a lot of internal state though...

Reply:
Hmm...  the files are created in 'java.io.tmpdir', but they'd be cleaned up by the sort, right?  If not, then one could check the number of files in that directory, etc.

Reply:
they are cleaned up on close - we could check the number of files in /tmp, but there's no guarantee that some other process isn't creating/deleting tmp files, so the counts would not be accurate.